api/v1alpha1/adminnetworkpolicy_types.go

```diff
@@ -0,0 +1,309 @@
+	// lower importance. An ANP with a priority of "0" will be evaluated after all
+	// positive priority AdminNetworkPolicies and standard NetworkPolicies.
+	// This field is NOT optional.
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=1000
+	Priority *int32 `json:"priority"`
+
+	// Subject defines the pods to which this AdminNetworkPolicy applies.
+	// This field is NOT optional.
+	Subject AdminNetworkPolicySubject `json:"subject"`
+
+	// List of Ingress rules to be applied to the selected objects.
+	// A total of 100 rules will be allowed per each network policy instance,
+	// this rule count will be calculated as the total summation of the
+	// Ingress and Egress rules in a single AdminNetworkPolicy Instance.
+	// If this field is empty then this AdminNetworkPolicy has no effect on
+	// ingress traffic.
+	// +optional
+	Ingress []AdminNetworkPolicyIngressRule `json:"ingress,omitempty"`
+
+	// List of Egress rules to be applied to the selected objects.
+	// A total of 100 rules will be allowed per each network policy instance,
+	// this rule count will be calculated as the total summation of the
+	// Ingress and Egress rules in a single AdminNetworkPolicy Instance.
+	// If this field is empty then this AdminNetworkPolicy has no effect on
+	// egress traffic.
+	// +optional
+	Egress []AdminNetworkPolicyEgressRule `json:"egress,omitempty"`
+}
+
+// AdminNetworkPolicyIngressRule describes an action to take on a particular
+// set of traffic destined for pods selected by an AdminNetworkPolicy's
+// Subject field. The traffic must match both ports and from.
+type AdminNetworkPolicyIngressRule struct {
+	// Name is an identifier for this rule, that should be no more than 100 characters
+	// in length.
+	// +optional
+	// +kubebuilder:validation:MaxLength=100
+	Name string `json:"name,omitempty"`
+
+	// Action specifies whether this rule must pass, allow or deny traffic.
+	// Allow: allows the selected traffic
+	// Deny: denies the selected traffic
+	// Pass: instructs the selected traffic to skip any remaining positive priority
+	// ANP rules, and then pass execution to any NetworkPolicies that select the pod.
+	// If the pod is not selected by any NetworkPolicies then execution
+	// is passed to any Priority=0 ANP rules that select the pod, whereupon the traffic
+	// is allowed if not dropped by any Priority=0 ANP rules.
+	// This field is mandatory.
+	Action AdminNetPolRuleAction `json:"action"`
+
+	// Ports allows for matching on traffic based on port and protocols.
+	// This field is mandatory.
+	Ports AdminNetworkPolicyPorts `json:"ports"`
+
+	// List of sources from which traffic will be allowed/denied/passed to the entities
+	// selected by this AdminNetworkPolicyRule. Items in this list are combined using a logical OR
+	// operation. If this field is empty, this rule matches no sources.
+	// If this field is present and contains at least one item, this rule
```

Comment:
The field is always present, because it is not optional.

What is the point of creating an AdminNetworkPolicyRule with an empty `From`? In NetworkPolicy, you need to be able to do that to create a "deny all" policy, since the only way to write that is "allow from none". But in ANP, a Rule with no Peers has absolutely no effect. Writing a Rule with no Peers is exactly the same as just _not writing that Rule_.

So either we should say that `From` is optional and a Rule with no `From` matches all sources, or else we should say that `From` is required to contain at least one item.

Reply:
Agreed I think it makes the most sense to just say `From` is required to contain at least one item.